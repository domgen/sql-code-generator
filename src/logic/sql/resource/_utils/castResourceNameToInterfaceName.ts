import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
