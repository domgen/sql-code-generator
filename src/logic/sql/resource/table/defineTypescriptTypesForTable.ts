import { pascalCase } from 'change-case';

import { TypeDefinitionOfResourceTable } from '../../../../model';

export const defineTypescriptTypesForTable = ({ definition }: { definition: TypeDefinitionOfResourceTable }) => {
  // define column types in typescript format
  const typescriptInterfaceColumnDefinitions = definition.columns.map((column) => {
    return `${column.name}: ${column.type.join(' | ')};`;
  });

  // output
  const typescriptInterfaceDefinition = `
export interface ${pascalCase(definition.name)} {
  ${typescriptInterfaceColumnDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return typescriptInterfaceDefinition;
};
