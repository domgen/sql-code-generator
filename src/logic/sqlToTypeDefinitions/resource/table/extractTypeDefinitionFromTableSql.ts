import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../common/castCommasInParensToPipesForTokenSafety';
import { extractTypeDefinitionFromColumnSql } from './extractTypeDefinitionFromColumnSql';
import { TypeDefinitionOfResourceTable } from '../../../../model';

/*
  note:
    - for table and view, just a map of { property: type }
    - for function and procedure, input type, output type
*/

const MYSQL_KEY_OR_CONSTRAINT_LINE_REGEX = /[KEY|key].*\s\(/g;

export const extractTypeDefinitionFromTableSql = ({ name, sql }: { name: string; sql: string }) => {
  // 0. strip comments
  const strippedSql: string = strip(sql);

  // 1. grab the insides of the "create" (i.e., 'CREATE TABLE ... (__INSIDES__) ...' => '__INSIDES__')
  const innerSqlAndAfter = strippedSql
    .split('(')
    .slice(1)
    .join('('); // drop the part before the first '('
  const innerSql = innerSqlAndAfter
    .split(')')
    .slice(0, -1)
    .join(')'); // drop the part after the last ')'

  // 2. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({ sql: innerSql });

  // 3. grab definition lines, by splitting out properties by commas
  const lines = parenTokenizedInnerSql.split(',');

  // 4. strip out the lines that are defining keys or constraints
  const columnDefinitions = lines
    .filter((line) => {
      return !new RegExp(MYSQL_KEY_OR_CONSTRAINT_LINE_REGEX, 'gm').test(line); // note, we reinstantiate so as to not share regex object (i.e., state) between calls
    })
    .map((line) => line.trim());

  // 5. get column definition from each property
  const columns = columnDefinitions.map((line) => extractTypeDefinitionFromColumnSql({ sql: line }));

  // 6. return the table definition
  return new TypeDefinitionOfResourceTable({
    name,
    columns,
  });
};
