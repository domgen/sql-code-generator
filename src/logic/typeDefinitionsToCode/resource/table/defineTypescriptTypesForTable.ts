import { TypeDefinitionOfResourceTable, ResourceType } from '../../../../model';
import { castResourceNameToTypescriptTypeName } from '../../common/castResourceNameToTypescriptTypeName';

export const defineTypescriptTypesForTable = ({ definition }: { definition: TypeDefinitionOfResourceTable }) => {
  // define column types in typescript format
  const typescriptInterfaceColumnDefinitions = definition.columns.map((column) => {
    return `${column.name}: ${column.type.join(' | ')};`;
  });

  // output
  const typescriptInterfaceDefinition = `
export interface ${castResourceNameToTypescriptTypeName({ name: definition.name, resourceType: ResourceType.TABLE })} {
  ${typescriptInterfaceColumnDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return typescriptInterfaceDefinition;
};
