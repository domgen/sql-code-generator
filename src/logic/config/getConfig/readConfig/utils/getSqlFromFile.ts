import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
