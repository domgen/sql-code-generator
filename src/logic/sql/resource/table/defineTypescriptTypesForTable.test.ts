import { defineTypescriptTypesForTable } from './defineTypescriptTypesForTable';
import { extractTypeDefinitionFromTableSql } from './extractTypeDefinitionFromTableSql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('defineTypescriptTypesForTable', () => {
  it('should generate an accurate looking interface for this table definition', async () => {
    const definition = extractTypeDefinitionFromTableSql({
      name: 'image',
      sql: await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/image.sql` }),
    });
    const code = defineTypescriptTypesForTable({ definition });
    expect(code).toMatchSnapshot();
  });
  it('should generate an accurate looking interface for this other table definition', async () => {
    const definition = extractTypeDefinitionFromTableSql({
      name: 'suggestion_version',
      sql: await getSqlFromFile({ filePath: `${__dirname}/__test_assets__/suggestion_version.sql` }),
    });
    const code = defineTypescriptTypesForTable({ definition });
    expect(code).toMatchSnapshot();
  });
});
