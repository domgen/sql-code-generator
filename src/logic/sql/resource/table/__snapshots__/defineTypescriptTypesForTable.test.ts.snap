// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForTable should generate an accurate looking interface for this other table definition 1`] = `
"export interface SuggestionVersion {
  id: number;
  suggestion_id: number;
  effective_at: Date;
  created_at: Date;
  status: string;
}"
`;

exports[`defineTypescriptTypesForTable should generate an accurate looking interface for this table definition 1`] = `
"export interface Image {
  id: number;
  uuid: string;
  created_at: Date;
  url: string;
  caption: string | null;
  credit: string | null;
  alt_text: string | null;
}"
`;
